### PR TITLE
Error marker when null analysis enabled

### LINKED ISSUE #854: Error marker when null analysis enabled
A related issue - https://github.com/redhat-developer/vscode-java/issues/3002

Steps to reproduce:
- import[ test project](https://github.com/snjeza/vscode-test/raw/master/testissue3002.zip)
- you will get this error
```
Type mismatch: cannot convert from Class<App.MyInit> to Class<? extends Init<? extends Configuration>>
```

### PR DESCRIPTION
**Pull request by snjeza**
Fixes #854

### CODE REVIEW

**Comment by snjeza** [comment]
@iloveeclipse It is similar to #629

**Comment by iloveeclipse** [comment]
> @iloveeclipse It is similar to #629

Thanks, but I really mean that a commit message should provide all essential information required to understand the change. It is later the only place where one can see explanation for a typical question "why on hell this line is doing this"?

**Comment by snjeza** [comment]
@iloveeclipse I have updated the PR.

**Comment by gayanper** [comment]
@snjeza You are providing lot of fixes to JDT, and sometimes its seems wasteful to repeat whats already there. But it usefull when we have a commit message which summaries the fix, specially if you are looking at git log from editor, cli or IDE which may or may not have github support.

I think what @iloveeclipse try describe is to follow this https://github.com/eclipse-jdt/eclipse.jdt.core/wiki/FAQ-JDT-Core-Committer#committing-and-creating-a-pr-for-review when you make the commit message. You can copy the same to the PR message a well. Personally even my self I fail time to time keep up with the format. But a quick look at the PR stack I can see https://github.com/eclipse-jdt/eclipse.jdt.core/pull/866/commits which follows the expected convention.

**Comment by snjeza** [comment]
@gayanper @iloveeclipse I have updated the PR - https://github.com/eclipse-jdt/eclipse.jdt.core/pull/855/commits/2dfa574065420b8f4c3838da906acb0c9787535b

**Review by stephan-herrmann — CHANGES_REQUESTED** [review]
I'm slightly worried about each of these tweaks affecting the order of resolution, but for lack of a strategy to avoid this (I'll think more about that later), I think the direction of this change is OK.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java [hunk 1]
```diff
@@ -1552,7 +1552,28 @@ private boolean isCompatibleWith0(TypeBinding otherType, /*@Nullable*/ Scope cap
...
 										// to a class is not allowed
 				return false;
-			return otherReferenceType.isSuperclassOf(this);
+			if (otherReferenceType.isSuperclassOf(this)) {
+				return true;
+			}
+			if (this instanceof SourceTypeBinding) {
+				SourceTypeBinding sourceType = (SourceTypeBinding) this;
+				if (sourceType.scope != null && sourceType.scope.referenceContext != null
+						&& sourceType.scope.compilerOptions().isAnnotationBasedNullAnalysisEnabled) {
```
[the remark this answers is not in the document]
Comment (snjeza) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java]: It has been fixed.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java [hunk 2]
```diff
@@ -1552,7 +1552,28 @@ private boolean isCompatibleWith0(TypeBinding otherType, /*@Nullable*/ Scope cap
...
+					TypeReference reference = sourceType.scope.referenceContext.superclass;
+					if (reference != null) {
+						if (!(reference.resolvedType instanceof ReferenceBinding)) {
+							reference.resolveType(sourceType.scope);
+						}
+						if (reference.resolvedType instanceof ReferenceBinding) {
+							ReferenceBinding binding = (ReferenceBinding) reference.resolvedType;
+							if (binding.isCompatibleWith(otherReferenceType)) {
```
[the remark this answers is not in the document]
Comment (snjeza) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java]: It has been fixed.

**Comment by snjeza** [comment]
@stephan-herrmann @iloveeclipse @gayanper I have updated the PR.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java [hunk 3]
```diff
@@ -1552,7 +1552,28 @@ private boolean isCompatibleWith0(TypeBinding otherType, /*@Nullable*/ Scope cap
...
 				return false;
-			return otherReferenceType.isSuperclassOf(this);
+			if (otherReferenceType.isSuperclassOf(this)) {
+				return true;
+			}
+			if (this instanceof SourceTypeBinding) {
+				SourceTypeBinding sourceType = (SourceTypeBinding) this;
+				if (sourceType.scope != null && sourceType.scope.referenceContext != null
+						&& sourceType.scope.compilerOptions().isAnnotationBasedNullAnalysisEnabled) {
+					TypeReference reference = sourceType.scope.referenceContext.superclass;
```
[the remark this answers is not in the document]
Comment (snjeza) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java:1569]: It has been fixed within https://github.com/eclipse-jdt/eclipse.jdt.core/issues/629

[the remark this answers is not in the document]
Comment (stephan-herrmann) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java:1569]: > It has been fixed within #629

I don't understand the connection. What has been fixed in #629 that affects `ReferenceBinding.isCompatible0()`?

[the remark this answers is not in the document]
Comment (snjeza) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java:1569]: #629 fixes a similar issue related to `superInterfaces`. Please see
- https://github.com/snjeza/eclipse.jdt.core/blob/master/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java#L1353
- https://github.com/snjeza/eclipse.jdt.core/blob/master/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTests18.java#L632
- https://github.com/snjeza/eclipse.jdt.core/blob/master/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTests18.java#L661
-

**Comment by snjeza** [comment]
The issue has been fixed via https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1067

**Comment by stephan-herrmann** [comment]
> The issue has been fixed via https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1067

After a 180 degree turn the issue will be fixed via #1077 instead of ~~#1067.~~

I'm sorry I couldn't give more substance to my vague worries up-front, which might have avoided some detours ...

**Pull request closed without merging**